lib/core/metadata.hpp

```diff
@@ -0,0 +1,147 @@
+    /// of the given length.  Returns the number of bytes consumed.
+    virtual size_t set_from_bytes(const char* /*bytes*/, size_t /*length*/) {
+        return 0;
+    }
+
+    /// Serializes this metadata object into the given byte array,
+    /// expected to be of length (at least) get_serialized_size().
+    virtual size_t serialize(char* /*bytes*/) {
+        return 0;
+    }
+
+    virtual nlohmann::json to_json();
+};
+
+void to_json(nlohmann::json& j, const metadataObject& m);
+void from_json(const nlohmann::json& j, metadataObject& m);
+
+CREATE_FACTORY(metadataObject);
+
+/**
+ * @brief A memory pool for preallocated metadata containers.
+ *
+ * The idea behind metadata containers is to be able to pass metadata down
+ * a pipeline chain between buffers and stages without copying any of the values
+ * at each step, or allocating and deallocating memory.
+ *
+ * This pool is what holds the metadata containers when they aren't in use
+ * and provides references to containers once they are requested.
+ *
+ * When the a metadata container's reference counter reaches zero, it returns
+ * itself back to its associated pool
+ *
+ * @author Andre Renard
+ */
+class metadataPool : public kotekan::kotekanLogging,
+                     public std::enable_shared_from_this<metadataPool> {
+    // see "Best" example in https://en.cppreference.com/w/cpp/memory/enable_shared_from_this
+    struct Private {};
+
+public:
+    // Constructor is only usable by this class
+    metadataPool(Private, int num_metadata_objects, size_t object_size,
+                 const std::string& unique_name, const std::string& type_name);
+    ~metadataPool();
+    // Everyone else has to use this factory function
+    // Hence all metadataPool objects are managed by shared_ptrs
+    static std::shared_ptr<metadataPool> create(int num_metadata_objects, size_t object_size,
+                                                const std::string& unique_name,
+                                                const std::string& type_name);
+
+    std::shared_ptr<metadataPool> get_shared() {
+        return shared_from_this();
+    }
+    std::weak_ptr<metadataPool> get_weak() {
+        return weak_from_this();
+    }
+
+    std::shared_ptr<metadataObject> request_metadata_object();
+
+    // void return_metadata_to_pool(struct metadataPool* pool, std::shared_ptr<metadataObject>
+    // container);
+
+    /// Name of the metadata pool
+    std::string unique_name;
+
+    /// Data type of the metadata objects in this pool
+    std::string type_name;
+
+    /// The size of the object stored in this pool
+    size_t metadata_object_size;
+
+protected:
+    /// The underlying block af data that we allocate objects out of
+    void* data_block;
+
+    /**
+     * @brief An array to indicate the use state of each pointer in the @c metadata_objects array
+     * A value of 1 indicates the pointer is in use and should have a reference count > 0
+     */
+    std::vector<bool> in_use;
```

Comment:
I don't know how this vector is accessed, but C++ has a quirk: All containers are thread-safe in the sense that different threads can access different elements simultaneously, as long as the container itself is not modified (grown or shrunk etc.). This guarantee does not hold for `std::vector<bool>`. If different threads access it simultaneously the result is undefined. One has to use `std::vector<char>` instead. I don't know whether this is the case here, but I'm bringing it up anyway just to be safe.

Reply:
Good catch!  It turns out that vector wasn't needed at all any more! :)